Xfoil interaction

AunSiro opened the pull request:
Bueno Colegas, me parece que este algoritmo ya está lo bastante pulido como para añadirlo al master! :D 
Si podéis, probadlo un poco y comprobad que va bien, y si todo está ok, propongo unirlo ya!

...O, quizás... ¿quizás esperar un poco más y terminar de documentar todos los scripts y funciones antes de fusionar?

AunSiro committed:
Primer Interfaz de comunicación mínimo

Compatible con Python 3

AunSiro committed:
Readme para Xfoil

AunSiro committed:
Notebook con creacion de perfiles con bezier

Añadido un notebook interactivo que explica cómo se construye el perfil
a partir del genoma.
Actualizado readme.

AunSiro committed:
Prototipo incompleto genetico

Incluye los módulos finales de traducción de genoma a puntos de perfil,
y comunicación de los perfiles con xfoil. Incluye 2 perfiles de ejemplo,
iguales entre sí.
No incluye bucle de optimización.

AunSiro committed:
Readme updated

AunSiro committed:
Version primitiva inicializador

Versión primitiva del iniciador de la población añadido.
Readme actualizado.

AunSiro committed:
Añadido testing al script de generación inicial

Ahora el rango de perfiles iniciales es mucho mayor, ya que los perfiles
son analizados para evitar inconsistencias.

AunSiro committed:
Algoritmo Funcional MK1

El algoritmo, aunque mejorable y con la documentación sin terminar, ya
se encuentra totalmente operativo.

AunSiro committed:
Funcional MK2

Ambiente Personalizable: Planeta, mach/velocidad, cuerda y altura de
vuelo.
Dibujador de perfiles
guardado automático de datos permite continuar desde un punto anterior.

AlexS12 commented:
¡Hola Siro!
Para mí una mejora buena sería si se pudiese ejecutar sin necesidad de meter el script en la carpeta en la que está Xfoil. Eso puede ser un poco rollo y te obliga a "mezclar instalaciones". yo creo que una forma fácil de solucionarlo podría ser que te pida el path donde está xfoil y lo busque ahí o algo así, ¿no?

Por lo demás muy chulo eh! me encanta!

astrojuanlu commented on aeropy/Xfoil_Interaction/Genetic_algorithm_files/ambient.py:
```diff
@@ -0,0 +1,247 @@
+    return conditions
+    
+    
+def temperatura(h, conditions, dT = 0):
```
Veo que estás rehaciendo la atmósfera estándar aquí :P ¿Hay algún motivo para ello? ¿En la versión actual te resulta complicado adaptar los coeficientes para Marte?

AunSiro replied:
Sí, rehago la atmósfera estándar! XD
La manera en la que lo estoy haciendo creo que es la que hace que sea más fácil tener distintos planetas, ya que las diferentes funciones atmosféricas no contienen datos en sí, sino que hay que alimentarlas con los que devuelve una función que contiene los datos. Así, para calcular diferentes planetas sólo cambia los datos con que alimentas las funciones atmosféricas.
No he usado la atmósfera más rápida de las vuestras porque sólo hay que calcularlo una vez, así que cogí la mía porque es la que mejor sé cómo funciona.

astrojuanlu replied:
¿Con este tipo de código piensas que estaría más fácil?

https://github.com/AeroPython/aeropy/blob/juanlu-numba/jnumba/isa.py

No lo hemos hablado todavía pero si esta es la más rápida estando en Python (#4) tal vez habría que poner la versión con numba. Si lo ves factible, intento pulirla un poco y completarla el domingo.

AunSiro replied:
Puedo intentar adaptar la versión con numba para que haga lo que quiero de comerse diferentes planetas. Pero añadiría un paquete de python más a los que hace falta tener instalados para poder ejecutar esto. Es lo bastante común?
 Al fin y al cabo, el tiempo que tarda en calcular una vez la atmósfera es totalmente despreciable con respecto al tiempo que tarda el xfoil en calcular un perfil...

astrojuanlu replied:
Sí, me imagino que el rendimiento no es el problema en este caso. Es un tema más filosófico que otra cosa, de esta forma intentamos reutilizar todos el código de todos, y este ha sido el primer caso donde se ha visto que haga falta :smile: 

¿A qué te refieres con instalar otro paquete? ¿Te refieres a numba en sí? ¿Es aumentar el número de dependencias lo que te preocupa?

AunSiro replied:
Sí, justo! Me imagino que vendrá con Anaconda, pero para la gente que se lo va instalando por su cuenta, será lo normal tenerlo? Le podría echar para atrás a alguien que lo necesite?
Quizás se puede detallar en la documentación luego dónde se usa cada paquete y si se puede prescindir de él comentando algunas líneas del código.

astrojuanlu replied:
Si están en Windows es muy muy raro que tengan NumPy y similares sin alguna distribución (Python(x,y), WinPython, Canopy o Anaconda). numba viene directamente instalado en Anaconda - fuera de ahí necesita un poco de trabajo, pero no más que instalar SciPy por ejemplo. En todo caso, se puede hacer que numba no sea obligatorio y meter un condicional: como solo es una línea, si hay numba estupendo y si no se queda como está.

astrojuanlu commented on aeropy/Xfoil_Interaction/Genetic_algorithm_files/ambient.py:
```diff
@@ -0,0 +1,247 @@
+        
+    
+    
+    re = Reynolds(densidad(height, planet_dic[planet]), chord, vel, viscosidad(temperatura(height, planet_dic[planet]), planet))
```
Nótese que esta línea se activa independientemente de si los datos son válidos o no. ¿Has pensado en lanzar un error más arriba?

astrojuanlu commented on aeropy/Xfoil_Interaction/Genetic_algorithm_files/analice.py:
```diff
@@ -0,0 +1,73 @@
+    
+def profile_analice (generation, profile_number):    
+    profile_name = 'gen' + str(generation) + 'prof' + str(profile_number)
+    data_root = "aerodata\data" + profile_name + '.txt'
```
Veo que estás metiendo rutas de Windows. Por si quieres que sea más portable y tal vez para facilitar lo que te dice Álex, te recomiendo que eches un vistazo a os.path:

http://pymotw.com/2/ospath/

astrojuanlu commented on aeropy/Xfoil_Interaction/Genetic_algorithm_files/testing.py:
```diff
@@ -0,0 +1,98 @@
+    test3 = prueba_creciente(perfil[50:75, 0])
+    test4 = prueba_creciente(perfil[75:100, 0])
+    
+    if (test1 * test2 * test3 * test4):
```
¿Qué estás probando aquí exactamente?

astrojuanlu replied:
Oh perdona, ya veo que esto es una función de ayuda.

AunSiro replied:
Hay 3 test que se ejecutan en orden de complejidad cada vez que se crea un nuevo perfil:
El primero comprueba una serie de relaciones básicas dentro de los parámetros que definen el perfil: p.ej, que el extradós queda por arriba del intradós, que los parámetros que son longitudes tienen un valor mínimo, etc. 
El segundo test, comprueba que los puntos que el perfil genera, que se calculan recorriendo el perfil desde el borde de salida en sentido antihorario, tienen siempre la cordenada x en sentido decreciente arriba y creciente abajo, para evitar que las curvas que forman el perfil tengan retrocesos y cosas raras.
El último test crea dos funciones interpolando los puntos del extrasós y el intrados, y comprueba que para 100 valores de x entre 0.1 y 0.9, la distancia entre ellos supera un mínimo, para evitar figuras extrañas en las que ambas líneas se crucen.

astrojuanlu commented:
¡Está genial! He puesto algunos comentarios pero por lo demás está bastante bien. Me descoloca un poco la mezcla de inglés y español en las variables y nombres de funciones, eso sí.

AunSiro commented:
Estoy corrigiendo y documentando lo de los nombres, tomo nota de todas las sugerencias!

Lo que no estoy muy seguro de si merece la pena es lo de buscar la ruta donde está Xfoil, ya que al menos en windows, se distribuye ya el exe listo, no hay que instalarlo.

AunSiro committed:
Funcional MK2.5

Unificadas funciones y variables en inglés, código correctamente
comentado.

AunSiro commented:
Ya he dejado el código en inglés y bien comentado!

Con respecto a lo de hacerlo multiplataforma, Los que usáis linux, me podéis decir cómo se abre el xfoil en linux? Me parece buena idea, sólo hay que cambiar un par de cosas en la manera en cómo se escriben las rutas.

Y con respecto a lo que ha dicho Álex? En Linux hay que instalarlo?

newlawrence commented:
Yo creo que sería mejor, para todo este tipo de programas externos, exigir
que se encuentren instalados y en el path y listo. El problema lo
tendríamos con aquellos que manejan ficheros en su propio directorio local,
que entonces tendríamos que preguntarle al SO desde dónde está ejecutando
el programa. En linux y Mac, tenemos el comando "wich" para eso; pero ahora
mismo no sé si en Windows hay algo parecido...

El 3 de marzo de 2015, 14:39, Siro notifications@github.com escribió:

> Ya he dejado el código en inglés y bien comentado!
> 
> Con respecto a lo de hacerlo multiplataforma, Los que usáis linux, me
> podéis decir cómo se abre el xfoil en linux? Me parece buena idea, sólo hay
> que cambiar un par de cosas en la manera en cómo se escriben las rutas.
> 
> Y con respecto a lo que ha dicho Álex? En Linux hay que instalarlo?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/AeroPython/aeropy/pull/20#issuecomment-76946257.

astrojuanlu commented:
En Linux en teoría se llama simplemente así:

```
$ xfoil
```

(el símbolo del dólar viene dado). ¿En Windows funciona sin el .exe? Porque en ese caso ya sería del todo portable. Luego es responsabilidad del usuario que el programa xfoil esté en el path, tanto en Windows como en Linux.

newlawrence commented:
Sí, en Windows funciona sin el .exe

El 3 de marzo de 2015, 18:48, Juan Luis Cano Rodríguez <
notifications@github.com> escribió:

> En Linux en teoría se llama simplemente así:
> 
> $ xfoil
> 
> (el símbolo del dólar viene dado). ¿En Windows funciona sin el .exe?
> Porque en ese caso ya sería del todo portable. Luego es responsabilidad del
> usuario que el programa xfoil esté en el path, tanto en Windows como en
> Linux.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/AeroPython/aeropy/pull/20#issuecomment-76996290.

AunSiro commented:
Acabo de comprobar que funciona sin el .exe, así que voy a cambiar cómo se escriben las líneas de carpetas y ya sería multiplataforma!

AunSiro committed:
Funcional MK2.6

Compatible con Linux

AunSiro commented:
En teoría ya debería ser multiplataforma!
Alguien con Linux se anima a probarlo? Lo que necesite del Xfoil se puede descargar de http://web.mit.edu/drela/Public/web/xfoil/

astrojuanlu commented:
No he instalado Xfoil aún, pero si corro el main me da un error:

```
$ python main.py 
Traceback (most recent call last):
  File "main.py", line 71, in <module>
    interfaz.xfoil_calculate_population(generation, ambient_data, aero_domain)
  File "/home/juanlu/Development/Python/AeroPython/aeropy/aeropy/Xfoil_Interaction/Genetic_algorithm_files/interfaz.py", line 106, in xfoil_calculate_population
    xfoil_calculate_profile(generation, profile_number, genome_matrix[profile_number-1,:], ambient_data, aero_domain)
  File "/home/juanlu/Development/Python/AeroPython/aeropy/aeropy/Xfoil_Interaction/Genetic_algorithm_files/interfaz.py", line 74, in xfoil_calculate_profile
    archivo = open(profile_root, mode = 'x')
FileNotFoundError: [Errno 2] No such file or directory: 'profiles/gen0/profile1.txt'
```

AunSiro committed:
Funcional MK2.7

Corregido bug de creación de carpetas

AunSiro commented:
Eliminé una cosa que no recordaba qué hacía. Mal hecho XD. Ya está corregido

astrojuanlu commented:
Lo he vuelto a probar de nuevo y no ha funcionado bien: falta un archivo `xfoil.def` al principio, no encuentra un perfil y a partir de ahí son todo "command not recognized".

https://gist.github.com/Juanlu001/42222736d6281982cdfe

AunSiro committed:
Funcional MK2.8

Corregido bug que impedía la lectura de los perfiles en el Xfoil de
Linux

AunSiro committed:
Funcional MK2.9

Añadidas opciones de final report, como imágenes y texto.

AunSiro committed:
Reorganizado el paquete

AunSiro committed:
Cambio a OOP

Creo que evitaré muchos bugs con una estrategia de programación
orientada a objetos
